### PR TITLE
Cleanup join / leave signal signatures

### PR DESCRIPTION
**Pull request by vladsud**
Splitting from https://github.com/microsoft/FluidFramework/pull/12042

ISignalClient was lying about its shape - extra properties exist only on ISignalMessage, even though they are applicable only for payloads of ISignalClient shape (i.e., join & leave signals).
I believes it would be actually more correct to have that payload on ISignalClient as it would ensure clients would get same info on connection for clients already in audience. But for now, I'm reflecting reality.

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts [hunk 1]
```diff
@@ -1798,10 +1798,15 @@ export class DeliLambda extends TypedEventEmitter<IDeliLambdaEvents> implements
      * @param signalMessage - Ticketed join signal message
      */
     private addSequencedSignalClient(clientJoinMessage: IClientJoin, signalMessage: ISignalMessageOutput) {
+        const op = signalMessage.message.operation;
+        assert(op.referenceSequenceNumber !== undefined,
+            "ISignalMessage.referenceSequenceNumber is undefined for sequenced signal");
+        assert(op.clientConnectionNumber !== undefined,
+            "ISignalMessage.clientConnectionNumber is undefined for sequenced signal");
```
Comment (vladsud) [inline server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: @GaryWilber, please take a look if this is correct (including services like FRS that do not fully support sequenced signals).

Overall feedback: I believe both fields should move from ISignalMessage to ISignalClient. Here are main reasons:
- Most important - client does not have that information when processing IConnected.initialClients. Without that data, it's impossible for clients to reason about any order here (even with this info it's really hard due to non-reliable signal delivery, but current structure makes it almost impossible to do anything).
- It would make some sense to keep it on ISignalMessage if other types of signals could receive this info. For example, if client could send some special signal that would roundtrip through Deli, such that it can always learn about "tip" (atomically increment and read clientConnectionNumber). But I'd guess we are not planning to have something like that, so having it on ISignalMessage  does not make much sense.

Somewhat related (building on last point) - not sure if it's good idea, but it might be useful to have a tool in the form of signal that client can send and receive back the following:
- copy of it's own join signal (i.e. learn about IClient, clientConnectionNumber, etc.)
- bump latest clientConnectionNumber and return current value.

This tool might be useful in the following cases:
- client never received its own join signal - currently client has to either reconnect or compromise its state machine (add more states and make its API more complex to use).
- As a conflict resolution tool (atomic ++ can be used by clients to build barriers, resolve conflicts, etc.)

Maybe these are two different tools (i.e. learning about own IClient may take different form, including asking for current state of Audience) and we add them on as-needed bases - not sure (need to start prototyping leader election using signals to have better confidence here)

Comment (GaryWilber) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: Here's how it shows up on the wire today:
image
image
Is the client unable to parse those numbers from initialClients?

Deli is using `ISignalMessage` since that's what `createRoomJoinMessage` and `createRoomLeaveMessage` return. The number properties on the interface are really only used server-side, so we would probably want to make a server-side only version of this interface and remove them from the client one.

Adding a tool like that sounds useful. Deli is keeping track each clients join signal data in `sequencedSignalClients`, so it should be able to respond to a request to fetch that information.
I don't think I would use a signal for the request part though. Frontends don't parse / have any smart logic relating to submitSignal. Since deli deals with ops, it would be easier to create a new op type that deli uses for processing these requests.

File: common/lib/protocol-definitions/src/clients.ts [hunk 2]
```diff
@@ -44,16 +44,6 @@ export interface ISignalClient {
...
-
-    /**
-     * Counts the number of signals sent by the client
-     */
-    clientConnectionNumber?: number;
-
-    /**
-     * Sequence number that indicates when the signal was created in relation to the delta stream
-     */
-    referenceSequenceNumber?: number;
```
Comment (GaryWilber) [inline common/lib/protocol-definitions/src/clients.ts:56]: Was removing these intentional?

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts (hunk 1 above)
Comment (GaryWilber) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: I re-read what I typed. I guess the "signals" messages sent to clients are still using `ISignalMessage`. Seems like we have a problem since there's 2 different interfaces used in 2 different spots? ISignalClient in initialClients and ISignalMessage in "signal"

File: common/lib/protocol-definitions/src/clients.ts (hunk 2 above)
Comment (anthony-murphy) [reply common/lib/protocol-definitions/src/clients.ts:56]: yeah. this looks wrong to me. the problem is how the runtime handles signals and strips the envelope

Comment (anthony-murphy) [reply common/lib/protocol-definitions/src/clients.ts:56]: the op and signal workload should be symmetrical, and is op the refSeq is on the outermost envelope. ideally like ops, the loader layer also adds this to outgoing signals as well

Comment (anthony-murphy) [reply common/lib/protocol-definitions/src/clients.ts:56]: ahh. thought this was ISignalMessage, but that is below, so this looks good

Comment (GaryWilber) [reply common/lib/protocol-definitions/src/clients.ts:56]: `initialClients` is `ISignalClient[]`. 
If this is removed, then the server cannot tell the connection/ref numbers for existing clients when a new client connects.

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts (hunk 1 above)
Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: what do we want the behavior to be? should initial clients just be the join signals? if so, we need to create a combined interface the server can send while we deprecate the existing format

File: common/lib/protocol-definitions/src/clients.ts (hunk 2 above)
Comment (anthony-murphy) [reply common/lib/protocol-definitions/src/clients.ts:56]: we probably want to keep this if want clients from initial clients to also have their ref seq, and connectionnumber

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts (hunk 1 above)
Comment (vladsud) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: Correct, it was documented in both places, but I do not believe it's provided in both places.
This PR attempts to reflect current reality (if I got it wrong - let me know).

As for what we want - I've spelled it out in original comment. If only join/leave signals will ever have that info, than that info should be on ISignalClient (inner content), as we want to get it as part of initial clients as well. If we even introduce other tools in our toolbox that may have same info, we will solve it then.

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: doesn't gary's screenshot above show it in both places?

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: it may be inconsistent, as i can't see the inner properties of the content, specifically the client, for the join signal case

Comment (GaryWilber) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: It's in both places now, so removing them from `ISignalClient` is incorrect.

The problem is that `initialClients` uses `ISignalClient`'s, the `join` signal sets `contents` to `ISignalMessage`, and `leave` signal sets `contents` to a client id (string).

I assume the client wants to be able to know the `clientConnectionNumber` and `referenceSequenceNumber` in all scenarios, which is why those properties are on both `ISignalClient` and `ISignalMessage`.
- `ISignalClient` for `initialClients`
- `ISignalMessage` for `join` and `leave` signals

Thoughts on what we could do:
1. Leave things as they are now and keep the properties on `ISignalClient` & `ISignalMessage` 
2. Change how `initialClients` or `join` / `leave` signals work and align them so they use the same format
   - We could set `clientConnectionNumber` and `referenceSequenceNumber` for the `ISignalClient` object that's in the `join` signal message? ([this](https://github.com/microsoft/FluidFramework/blob/d8f825ec284baf904912a2cbb82c87164ed463e8/server/routerlicious/packages/lambdas/src/utils/messageGenerator.ts#L31))
   - But if we do that, then there's a mismatch with the leave signal ([this](https://github.com/microsoft/FluidFramework/blob/d8f825ec284baf904912a2cbb82c87164ed463e8/server/routerlicious/packages/lambdas/src/utils/messageGenerator.ts#L48)). The content is set to a string.

Basically `initialClients`, `join` signals, and `leave` signals all work differently and it's a mess 😞

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: we don't use any of those properties anywhere in the client yet, so we have some flexibility. we should figure out how it should be, and then make it happen

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: going back to my other comment, i think we should aim for symmetry between ops and signals, so i think from that we need clientConnectionNumber and referenceSequenceNumber on the out envelope, so ISignalMessage. Initial clients is more like a summary format, and like the summary format of the quorum, it should contain the reference sequence number the client connected at. it doesn't need clientConnectionNumber as that is really only applicable to signals i think

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: if you look at how it works for quorum, the consumer combines that out envelope with the client to create the actual member object:
https://github.com/microsoft/FluidFramework/blob/4272d4b6256634f5a2e35625bad46e3e51cc3391/server/routerlicious/packages/protocol-base/src/protocol.ts#L128

Comment (vladsud) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: Today none of this info is used, so we can change it however we want it.

Here are couple rules that would be nice to hit (looking at it from usability POV):

1. (pri1) How client processes join signals & initialClients should be the same. I.e., shape of data that client works in both cases should be the same. See https://github.com/microsoft/FluidFramework/pull/12042/files#diff-3446c9530009bbd303e61d96258d5cd1e75035e0baae2fbfffd29653de8c2603R750 for more details 
   - this is not a strong rule, as long as we can convert initialClients info to appropriate shape to look like a signal we just got, we should be fine, but ideally that's trivial operation.
2. (pri2) There is global order (exposed to all clients) for all join / leave signals. I.e. each client can say that client A left audience before client B joined, even if signals are reordered.
   - This would be strong requirement for building any leader election using signals IF signals were reliable. Given that signals are not reliable, it's possible it does not matter as clients will have to rely on continuous polling of some sort to be able to have any kind of leader election using signals - we will learn by implementing a prototype here.
   - That said, any piece of info helps. So, exposing clientConnectionNumber & referenceSequenceNumber (and ideally in the same shape as with join signal) would be appreceated.

Comment (vladsud) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: As for Tony's suggesting (promoting type and maybe not relying on signal.clientId === null to differentiate system vs. non-system signals) - if feels a bit orthogonal and optional to me. It would be nice to have but having slightly different schema is not the end of the world. IF it helps to get to final state faster - sure, but if it helps to expose join info in reasonable shape faster (and make that changes later) - that would work too

Comment (anthony-murphy) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:1805]: on caveat to that is that we do use initial client today, so we can't break that usage, so only additive changes can happen there.

i don't get the signal.clientId === null reference. My preference is the outer most envelope of all signals has referenceSequenceNumber eventually, just like ops,  in join leave the server sets in, for client scenarios the client sets it, just like ops.
